internal/release/observed.go

```diff
@@ -0,0 +1,200 @@
+	"encoding/json"
+	"io"
+
+	"github.com/mitchellh/copystructure"
+	"helm.sh/helm/v3/pkg/chart"
+	helmrelease "helm.sh/helm/v3/pkg/release"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	helmv2 "github.com/fluxcd/helm-controller/api/v2beta2"
+	"github.com/fluxcd/helm-controller/internal/chartutil"
+	"github.com/fluxcd/helm-controller/internal/digest"
+)
+
+var (
+	DefaultDataFilters = []DataFilter{
+		IgnoreHookTestEvents,
+	}
+)
+
+// DataFilter allows for filtering data from the returned ObservedRelease while
+// making an observation.
+type DataFilter func(rel *ObservedRelease)
+
+// IgnoreHookTestEvents ignores test event hooks. This allows to exclude it
+// while e.g. generating a digest for the object. To prevent manual test
+// triggers from a user to interfere with the checksum.
+func IgnoreHookTestEvents(rel *ObservedRelease) {
+	if len(rel.Hooks) > 0 {
+		var hooks []helmrelease.Hook
+		for _, h := range rel.Hooks {
+			if !IsHookForEvent(&h, helmrelease.HookTest) {
+				hooks = append(hooks, h)
+			}
+		}
+		rel.Hooks = hooks
+	}
+}
+
+// ObservedRelease is a copy of a Helm release object, as observed to be written
+// to the storage by a storage.Observer. The object is detached from the Helm
+// storage object, and mutations to it do not change the underlying release
+// object.
+type ObservedRelease struct {
+	// Name of the release.
+	Name string `json:"name"`
+	// Version of the release, at times also called revision.
+	Version int `json:"version"`
+	// Info provides information about the release.
+	Info helmrelease.Info `json:"info"`
+	// ChartMetadata contains the current Chartfile data of the release.
+	ChartMetadata chart.Metadata `json:"chartMetadata"`
+	// Config is the set of extra Values added to the chart.
+	// These values override the default values inside the chart.
+	Config map[string]interface{} `json:"config"`
+	// Manifest is the string representation of the rendered template.
+	// It is omitted from the JSON representation of the object for performance
+	// reasons. ManifestChecksum can be used to notice changes, after which
+	// Manifest can be inspected while loaded.
+	Manifest string `json:"manifest"`
+	// Hooks are all the hooks declared for this release, and the current
+	// state they are in.
+	Hooks []helmrelease.Hook `json:"hooks"`
+	// Namespace is the Kubernetes namespace of the release.
+	Namespace string `json:"namespace"`
+	// Labels of the release.
+	Labels map[string]string `json:"labels"`
+}
+
+// Targets returns if the release matches the given name, namespace and
+// version. If the version is 0, it matches any version.
+func (o ObservedRelease) Targets(name, namespace string, version int) bool {
+	return o.Name == name && o.Namespace == namespace && (version == 0 || o.Version == version)
+}
+
+// Encode JSON encodes the ObservedRelease and writes it into the given writer.
+func (o ObservedRelease) Encode(w io.Writer) error {
+	enc := json.NewEncoder(w)
+	if err := enc.Encode(o); err != nil {
```

Comment:
It is stated above that the `Manifest` is omitted in JSON representation, but this encode the whole object.

Reply:
See https://github.com/fluxcd/helm-controller/pull/503#discussion_r914676584